Bio.PDB sorting not tested and Python 2 only

The ``cmp`` function is Python 2 only, sorting using a key function is the preferred idiom for Python 2 and 3.

``Bio/PDB/Chain.py`` class ``Chain`` uses ``cmp`` in never used private method ``._sort```, could remove this?

``Bio/PDB/Structure.py`` class ``Structure``, same.

``Bio/PDB/Model.py`` class ``Model``, same.

``Bio/PDB/Residue.py`` class ``Residue`` defines ``._sort`` and uses in public ``.sort`` but not tested.

Spotted via flake8 tests (installed and running under Python 3),

```
$ flake8 --config=/dev/null --select F821 Bio/PDB/
Bio/PDB/Chain.py:34:20: F821 undefined name 'cmp'
Bio/PDB/Chain.py:36:20: F821 undefined name 'cmp'
Bio/PDB/Chain.py:37:16: F821 undefined name 'cmp'
Bio/PDB/Model.py:52:16: F821 undefined name 'cmp'
Bio/PDB/Residue.py:52:20: F821 undefined name 'cmp'
Bio/PDB/Residue.py:62:20: F821 undefined name 'cmp'
Bio/PDB/Residue.py:67:16: F821 undefined name 'cmp'
Bio/PDB/Structure.py:36:16: F821 undefined name 'cmp'
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -19,6 +19,14 @@ cache: pip
 matrix:
   include:
     - stage: basics
+      python: 2.7
+      env: TOXENV=style
+      addons:
+        apt:
+          packages:
+      before_install: echo "Going to run basic checks"
+    - stage: basics
+      python: 3.6
       env: TOXENV=style
       addons:
         apt:

--- a/Bio/PDB/Chain.py
+++ b/Bio/PDB/Chain.py
@@ -13,30 +13,6 @@ class Chain(Entity):
         self.level = "C"
         Entity.__init__(self, id)
 
-    # Private methods
-
-    def _sort(self, r1, r2):
-        """Sort function for residues in a chain (PRIVATE).
-
-        Residues are first sorted according to their hetatm records.
-        Protein and nucleic acid residues first, hetatm residues next,
-        and waters last. Within each group, the residues are sorted according
-        to their resseq's (sequence identifiers). Finally, residues with the
-        same resseq's are sorted according to icode.
-
-        Arguments:
-
-        - r1, r2 - Residue objects
-
-        """
-        hetflag1, resseq1, icode1 = r1.id
-        hetflag2, resseq2, icode2 = r2.id
-        if hetflag1 != hetflag2:
-            return cmp(hetflag1[0], hetflag2[0])
-        elif resseq1 != resseq2:
-            return cmp(resseq1, resseq2)
-        return cmp(icode1, icode2)
-
     def _translate_id(self, id):
         """Translate sequence identifer to tuple form (PRIVATE).
 
@@ -55,8 +31,6 @@ class Chain(Entity):
         if isinstance(id, int):
             id = (' ', id, ' ')
         return id
-
-    # Special methods
 
     def __getitem__(self, id):
         """Return the residue with given id.

--- a/Bio/PDB/Model.py
+++ b/Bio/PDB/Model.py
@@ -31,34 +31,8 @@ class Model(Entity):
 
         Entity.__init__(self, id)
 
-    # Private methods
-
-    def _sort(self, c1, c2):
-        """Sort the Chains instances in the Model instance.
-
-        Chain instances are sorted alphabetically according to
-        their chain id. Blank chains come last, as they often consist
-        of waters.
-
-        Arguments:
-         - c1, c2 - Chain objects
-
-        """
-        id1 = c1.get_id()
-        id2 = c2.get_id()
-        # make sure blank chains come last (often waters)
-        if id1 == " " and not id2 == " ":
-            return 1
-        elif id2 == " " and not id1 == " ":
-            return -1
-        return cmp(id1, id2)
-
-    # Special methods
-
     def __repr__(self):
         return "<Model id=%s>" % self.get_id()
-
-    # Public
 
     def get_chains(self):
         for c in self:

--- a/Bio/PDB/Residue.py
+++ b/Bio/PDB/Residue.py
@@ -29,47 +29,11 @@ class Residue(Entity):
         self.segid = segid
         Entity.__init__(self, id)
 
-    # Special methods
-
     def __repr__(self):
         resname = self.get_resname()
         hetflag, resseq, icode = self.get_id()
         full_id = (resname, hetflag, resseq, icode)
         return "<Residue %s het=%s resseq=%s icode=%s>" % full_id
-
-    # Private methods
-
-    def _sort(self, a1, a2):
-        """Sort the Atom objects.
-
-        Atoms are sorted alphabetically according to their name,
-        but N, CA, C, O always come first.
-
-        Arguments:
-         - a1, a2 - Atom objects
-
-        """
-        name1 = a1.get_name()
-        name2 = a2.get_name()
-        if name1 == name2:
-            return cmp(a1.get_altloc(), a2.get_altloc())
-        if name1 in _atom_name_dict:
-            index1 = _atom_name_dict[name1]
-        else:
-            index1 = None
-        if name2 in _atom_name_dict:
-            index2 = _atom_name_dict[name2]
-        else:
-            index2 = None
-        if index1 and index2:
-            return cmp(index1, index2)
-        if index1:
-            return -1
-        if index2:
-            return 1
-        return cmp(name1, name2)
-
-    # Public methods
 
     def add(self, atom):
         """Add an Atom object.

--- a/Bio/PDB/Residue.py
+++ b/Bio/PDB/Residue.py
@@ -84,7 +84,25 @@ class Residue(Entity):
         Entity.add(self, atom)
 
     def sort(self):
-        self.child_list.sort(self._sort)
+        """Sort child atoms.
+
+        Atoms N, CA, C, O always come first, thereafter alphabetically
+        by name, with any alternative location specifier for disordered
+        atoms (altloc) as a tie-breaker.
+        """
+        # Defining sort key function within the sort method's scope:
+        def sort_index(atom):
+            """Build tuple of (int, name, alt-loc) for sorting.
+
+            The first integer is 0, 1, 2, 3, 4 for atoms N, CA, C, O, other.
+            """
+            try:
+                i = ["N", "CA", "C", "O"].index(atom.name)
+            except ValueError:
+                i = 4
+            return (i, atom.name, atom.altloc)
+
+        self.child_list.sort(key=sort_index)
 
     def flag_disordered(self):
         """Set the disordered flag."""

--- a/Bio/PDB/Structure.py
+++ b/Bio/PDB/Structure.py
@@ -15,27 +15,8 @@ class Structure(Entity):
         self.level = "S"
         Entity.__init__(self, id)
 
-    # Special methods
-
     def __repr__(self):
         return "<Structure id=%s>" % self.get_id()
-
-    # Private methods
-
-    def _sort(self, m1, m2):
-        """Sort models.
-
-        This sorting function sorts the Model instances in the Structure instance.
-        The sorting is done based on the model id, which is a simple int that
-        reflects the order of the models in the PDB file.
-
-        Arguments:
-         - m1, m2 - Model instances
-
-        """
-        return cmp(m1.get_id(), m2.get_id())
-
-    # Public
 
     def get_models(self):
         for m in self:

--- a/Tests/test_PDB.py
+++ b/Tests/test_PDB.py
@@ -530,11 +530,27 @@ class ParseTest(unittest.TestCase):
             os.remove(filename)
 
     def test_deepcopy_of_structure_with_disorder(self):
-        """Test deepcopy of a structure with disordered atoms.
+        """Test deepcopy of a structure with disordered atoms; and sort it.
 
         Shouldn't cause recursion.
         """
-        _ = deepcopy(self.structure)
+        structure = deepcopy(self.structure)
+        # Sorting a residue modifies it in place, but since we
+        # have made a copy this won't alter the other tests
+        # using self.structure
+        for residue in structure.get_residues():
+            old = [a.name for a in residue]
+            residue.sort()
+            new = [a.name for a in residue]
+            special = []
+            for a in ['N', 'CA', 'C', 'O']:
+                if a in old:
+                    special.append(a)
+            special_len = len(special)
+            self.assertEqual(new[0:special_len], special,
+                             "Sorted residue did not place N, CA, C, O first: %s" % new)
+            self.assertEqual(new[special_len:], sorted(new[special_len:]),
+                                 "After N, CA, C, O should be alphabet: %s" % new)
 
 
 class ParseReal(unittest.TestCase):
@@ -551,6 +567,24 @@ class ParseReal(unittest.TestCase):
             self.assertFalse(len(struct))
         finally:
             os.remove(filename)
+
+    def test_residue_sort(self):
+        """Sorting atoms in residues."""
+        parser = PDBParser(PERMISSIVE=False)
+        structure = parser.get_structure("example", "PDB/1A8O.pdb")
+        for residue in structure.get_residues():
+            old = [a.name for a in residue]
+            residue.sort()
+            new = [a.name for a in residue]
+            special = []
+            for a in ['N', 'CA', 'C', 'O']:
+                if a in old:
+                    special.append(a)
+            special_len = len(special)
+            self.assertEqual(new[0:special_len], special,
+                             "Sorted residue did not place N, CA, C, O first: %s" % new)
+            self.assertEqual(new[special_len:], sorted(new[special_len:]),
+                                 "After N, CA, C, O should be alphabet: %s" % new)
 
     def test_c_n(self):
         """Extract polypeptides from 1A80."""

